dapps-certification-persistence/src/IOHK/Certification/Persistence/API.hs

```diff
@@ -1,16 +1,89 @@
+{-# LANGUAGE OverloadedLabels    #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RecordWildCards     #-}
 {-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE OverloadedLabels #-}
-{-# LANGUAGE RecordWildCards #-}
-{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeOperators       #-}
 
 module IOHK.Certification.Persistence.API where
 
-import Database.Selda
-import Database.Selda.SQLite
-import IOHK.Certification.Persistence.Structure
-import Data.Maybe
-import Control.Monad
+import           Control.Monad
+import           Data.Maybe
+import           Database.Selda
+import           Database.Selda.SQLite
+import           IOHK.Certification.Persistence.Structure
+
+getTransactionIdQ:: Text -> Query t (Col t (ID Transaction))
+getTransactionIdQ  externalAddress = do
+  p <- select transactions
+  restrict (p ! #wtxExternalId .== text externalAddress)
+  pure (p ! #wtxId)
+
+getTransactionId :: MonadSelda m => Text -> m (Maybe (ID Transaction))
+getTransactionId = fmap listToMaybe . query . getTransactionIdQ
+
+-- | inserts or updates a transaction
+-- NOTE: if the transaction exists Nothing will be returned and entries
+-- will not be updated
+upsertTransaction :: (MonadSelda m,MonadMask m)
+                  => Transaction
+                  -> [TransactionEntry]
+                  -> m (Maybe (ID Transaction))
+upsertTransaction tx@Transaction{..} entries = do
+  txIdM <- upsert transactions
+     (\p -> p ! #wtxExternalId .== literal wtxExternalId)
+     (`with`
+      [ #wtxExternalId := literal wtxExternalId
+      , #wtxAmount := literal wtxAmount
+      , #wtxTime := literal wtxTime
+      , #wtxDepth := literal wtxDepth
+      , #wtxStatus := literal wtxStatus
+      , #wtxMetadata := literal wtxMetadata
+      ])
+     [tx { wtxId = def :: ID Transaction}]
+  -- if transaction was just inserted , insert entries as well
+  -- otherwise we don't care to update entries
```

Comment:
Why not?

Reply:
It is just an optimization because a transaction can change its status or time stamps (maybe) but not the inputs or outputs (which are represented here by entries). 
Maybe I should add this explanation there too, no?